pkg/autoscaler/statforwarder/forwarder.go

```diff
@@ -0,0 +1,256 @@
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	leaseinformer "knative.dev/pkg/client/injection/kube/informers/coordination/v1/lease"
+	endpointsinformer "knative.dev/pkg/client/injection/kube/informers/core/v1/endpoints"
+	serviceinformer "knative.dev/pkg/client/injection/kube/informers/core/v1/service"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/hash"
+	"knative.dev/pkg/reconciler"
+	"knative.dev/pkg/system"
+)
+
+const (
+	// The port on which autoscaler WebSocket server listens.
+	autoscalerPort     = 8080
+	autoscalerPortName = "http"
+)
+
+// Forwarder dose the following things:
+// 1. Watches the change of Leases for Autoscaler buckets. Stores the
+//    Lease -> IP mapping.
+// 2. Creates/updates the corresponding K8S Service and Endpoints.
+// 3. Can be used to forward the metrics falling in a bucket based on
+//    the holder IP. (This is a TODO)
+type Forwarder struct {
+	selfIP          string
+	logger          *zap.SugaredLogger
+	kc              kubernetes.Interface
+	serviceLister   corev1listers.ServiceLister
+	endpointsLister corev1listers.EndpointsLister
+	// bs is the BucketSet including all Autoscaler buckets.
+	bs *hash.BucketSet
+	// Lock for leaseHolders.
+	leaselock sync.RWMutex
+	// leaseHolders stores the Lease -> IP relationships.
+	leaseHolders map[string]string
+}
+
+// New creates a new Forwarder.
+func New(ctx context.Context, logger *zap.SugaredLogger, kc kubernetes.Interface, selfIP string, bs *hash.BucketSet) *Forwarder {
+	ns := system.Namespace()
+	bkts := bs.Buckets()
+	endpointsInformer := endpointsinformer.Get(ctx)
+	f := Forwarder{
+		selfIP:          selfIP,
+		logger:          logger,
+		kc:              kc,
+		serviceLister:   serviceinformer.Get(ctx).Lister(),
+		endpointsLister: endpointsInformer.Lister(),
+		bs:              bs,
+		leaseHolders:    make(map[string]string, len(bkts)),
+	}
+
+	leaseInformer := leaseinformer.Get(ctx)
+	leaseInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: reconciler.NamespaceFilterFunc(ns),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    f.leaseUpdated,
+			UpdateFunc: controller.PassNew(f.leaseUpdated),
+			// TODO(yanweiguo): Set up DeleteFunc.
+		},
+	})
+
+	return &f
+}
+
+func (f *Forwarder) leaseUpdated(obj interface{}) {
+
+	l := obj.(*coordinationv1.Lease)
+	ns, n := l.Namespace, l.Name
+
+	if !f.bs.HasBucket(n) {
+		// Skip this Lease as it is not for Autoscaler.
+		return
+	}
+
+	identity := *l.Spec.HolderIdentity
+	if identity == "" {
+		f.logger.Debugf("HolderIdentity not found for Lease %s/%s", ns, n)
+		return
+	}
+
+	if h, ok := f.getHolder(n); ok {
+		if h == identity {
+			// Already up-to-date.
+			return
+		}
+	}
```

Comment:
I think I'd do this in one `if`, but this is personal preference, if you prefer it nested cool with me

```suggestion
	if h, ok := f.getHolder(n); ok && h == identity {
		// Already up-to-date.
		return
	}
```